### PR TITLE
Snyk reader

### PR DESCRIPTION
**Pull request by 0xgoto**

### CODE REVIEW

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/Reader.java [hunk 1]
```diff
@@ -96,7 +96,8 @@ public static List<Reader> allReaders() {
                 new WebInspectReader(),
                 new XanitizerReader(),
                 new ZapJsonReader(),
-                new ZapReader());
+                new ZapReader(),
+                new SnykReader());
```
Comment (darkspirit510) [inline plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/Reader.java:100]: For my "OCD", please order the readers alphabetically.

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
+package org.owasp.benchmarkutils.score.parsers;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.owasp.benchmarkutils.score.*;
```
Comment (darkspirit510) [inline plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java:7]: Please change wildcard import to concrete ones.

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+public class SnykReader extends Reader {
```
Comment (darkspirit510) [inline plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java:9]: Can you please copy-pate the usual header from another reader?

File: plugin/src/test/java/org/owasp/benchmarkutils/score/parsers/SnykReaderTest.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package org.owasp.benchmarkutils.score.parsers;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.owasp.benchmarkutils.score.*;
+
+public class SnykReaderTest extends ReaderTestBase {
```
Comment (darkspirit510) [inline plugin/src/test/java/org/owasp/benchmarkutils/score/parsers/SnykReaderTest.java:10]: Can you please copy-paste the usual header from another test header?

File: plugin/src/test/java/org/owasp/benchmarkutils/score/parsers/SnykReaderTest.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
+package org.owasp.benchmarkutils.score.parsers;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.owasp.benchmarkutils.score.*;
```
Comment (darkspirit510) [inline plugin/src/test/java/org/owasp/benchmarkutils/score/parsers/SnykReaderTest.java:8]: Please change wildcard import to concrete ones.

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                TestCaseResult tcr = new TestCaseResult();
+
+                String ruleId = result.getString("ruleId");
+                ruleId = (ruleId.substring(ruleId.lastIndexOf('/') + 1)).split("\\.")[0];
+
+                int cwe = snykCweMap.getOrDefault(ruleId, INVALID_RULE_ID);
+
+                if (cwe == INVALID_RULE_ID) {
+                    System.out.println("CWE # not parseable from: " + ruleId);
+                    return tcr;
```
Comment (darkspirit510) [inline plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java:79]: If the CWE number couldn't be read, I guess you should just return null. Right @davewichers ?

**Review by darkspirit510 — CHANGES_REQUESTED** [review]

**Comment by darkspirit510** [comment]
Thank you for this PR. I wrote some comments to the files. Can you somehow reduce the size of the test file?

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java (hunk 6 above)
Comment (davewichers) [reply plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java:79]: I believe that's right. If null is returned the rest case result is not added to the results set, which is what you want. You should test it to make sure it doesn't throw an NPE, but I think that's right.

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java (hunk 2 above)
Comment (0xgoto) [reply plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java:7]: Intellij is replacing the headers with a wildcard.

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        } catch (Exception ex) {
+            ex.printStackTrace();
+        }
+
+        return null;
+    }
+
+    private Boolean isSnyk(ResultFile resultFile) {
+
+        return resultFile.json().has("runs")
```
Comment (darkspirit510) [inline plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java:105]: One additional thing that came up to my mind this morning. On your path to the string `SnykCode` you use `has` method to check some, but not all parts of the structure. This still might throw an JSON exception. I got around this by using a try-catch and not check for the path at all (see `HorusecReader.canRead()`). Maybe you can just use a try-catch instead?

File: plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java (hunk 6 above)
Comment (0xgoto) [reply plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java:83]: What if a new ruleId is added to snyk code?
If a new ruleId is added, then the tool will always throw NullPointerException. So, i felt that adding an empty testCaseResult is a better approach. 

This is my opinion. What do you folks think? @darkspirit510 @davewichers

Comment (darkspirit510) [reply plugin/src/main/java/org/owasp/benchmarkutils/score/parsers/SnykReader.java:84]: Did you try this? Other readers check for null return and skip adding it to the collection 🤔

**Comment by 0xgoto** [comment]
I have resolved everything that @darkspirit510 stated.

**Comment by davewichers** [comment]
@darkspirit510 - can you review his last commit and confirm everything is good from your perspective before I review (and hopefully merge)?

**Pull request merged**